sympy/assumptions/assume.py

```diff
@@ -211,6 +307,131 @@ def eval(self, expr, assumptions=True):
         return res
 
 
+class AskHandlerClass(Dispatcher):
+    """
+    Class for handler which evaluates the ``AppliedPredicate``.
+
+    Explanation
+    ===========
+
+    This class dispatches various types, and return fuzzy boolean values
+    when called.
+    When the dispatched function return ``None``, the next function for
+    the signature is queried until some other value is returned, or no
+    more registered function is left.
+
+    Parameters
+    ==========
+
+    name : str
+
+    doc : str, optional
+
+    base : tuple of AskHandlerClass, optional
+        Base dispatchers, which inherites the dispatched functions.
+
+    Examples
+    ========
+
+    Generate a base dispatcher ``MyHandler``.
+
+    >>> from sympy import Symbol, Add, AskHandlerClass
+    >>> from sympy.abc import x
+    >>> MyHandler = AskHandlerClass('myhandler')
+    >>> @MyHandler.register(Symbol)
+    ... def _(expr, assumptions):
+    ...     return True
+    >>> @MyHandler.register(Add)
+    ... def _(expr, assumptions):
+    ...     return False
+    >>> MyHandler(x, assumptions=True)
+    True
+    >>> MyHandler(x+1, assumptions=True)
+    False
+    >>> print(MyHandler(2*x, assumptions=True)) # unregistred type
+    None
+
+    Generate new dispatcher ``MyHandler2``, which uses ``MyHandler`` as
+    base dispatcher
+
+    >>> MyHandler2 = AskHandlerClass('myhandler2', base=MyHandler)
+    >>> @MyHandler2.register(Add)
+    ... def _(expr, assumptions):
+    ...     return True
+    >>> MyHandler2(x, assumptions=True) # refer to base
+    True
+    >>> MyHandler2(x+1, assumptions=True)
+    True
+
+    """
+
+    def __init__(self, name, doc=None, base=()):
+        super().__init__(name, doc)
+        if not type(base) is tuple:
+            base = (base,)
+        for b in reversed(base):
+            self.funcs.update(b.funcs)
+        self.ordering = ordering(self.funcs)
+        self.base = base
+
+    def __call__(self, *args, **kwargs):
+        # If `None` is returned, search for next function
+        # If all signature return `None`, return `None`
+        types = tuple([type(arg) for arg in args])
+
+        try:
+            func = self._cache[types]
+        except KeyError:
+            func = self.dispatch(*types)
+            self._cache[types] = func
+
+        if not func:
+            return None
+
+        try:
+            # MDNotImplementedError may be raised here...
+            result = func(*args, **kwargs)
+            if result is None:
+                # ...or here
+                raise MDNotImplementedError
+            return result
+
+        except MDNotImplementedError:
```

Comment:
I always dislike this pattern of raising the exception just to catch it immediately. I think there's always a simpler way. In this case it's:
```python
        try:
            result = func(*args, **kwargs)
        except MDNotImplementedError:
            pass
        else:
            if result is not None:
                return result
```
With less nesting it's:
```python
        result = None
        try:
            # might return None
            result = func(*args, **kwargs)
        except MDNotImplementedError:
            pass
        if result is not None:
            return result
```